feat(frontend/form.io): Create custom provider that works with MapTiler

JumpLink opened the pull request:
This PR implements a custom MapTiler address provider for Form.io and restricts the form builder to use only our supported providers. The changes ensure that all form components use the correct providers and prevent errors from unsupported provider configurations.

### Key Changes

1. **MapTiler Address Provider Implementation**
   - Created `BaseAddressProvider` abstract class for extensible provider architecture
   - Implemented `MapTilerAddressProvider` with full geocoding support
   - Provider integrates with existing MapTiler client and supports international addresses
   - Results are transformed to Form.io-compatible format for seamless integration

2. **Provider Restriction & Default Configuration**
   - Restricted form builder to only enable MapTiler (address) and Beabee (storage) providers
   - Set MapTiler as default address provider for all address components
   - Set Beabee storage as default for all file upload components
   - Hidden provider/storage selection UI elements to prevent confusion

3. **Form Builder Configuration Updates**
   - Updated `formOpts` to automatically set correct providers for address and file components
   - Added CSS rules to hide provider/storage selection fields
   - Ensured all new forms use the correct provider configuration

4. **Database Migration for Existing Forms**
   - Created migration `1756466495968-UpdateCalloutFormSchemaComponents` to update existing callout forms
   - Automatically sets `storage: "beabee"` for all file components
   - Automatically sets `provider: "maptiler"` for all address components
   - Ensures backward compatibility for forms created before this change

5. **Type Safety & Architecture**
   - Added comprehensive TypeScript interfaces for provider options and results
   - Created extensible base classes for future provider implementations
   - Maintained backward compatibility with existing Form.io address component structure

6. **Dev CLI Fix**
   - Fixed `generate:index` script in frontend package.json (removed `start` command)

7. **Address Formatter & Testing**
   - Created `AddressFormatter` class for unified address data handling
   - Added comprehensive tests using Vitest framework
   - Supports pattern-based address formatting with fallback placeholders
   - Includes real-world test data from CERN, NASA, and Stonehenge
   - Transforms between MapTiler, Form.io, and unified address formats

8. **Router & Navigation Updates**
   - Updated route names in `CalloutMapHeader.vue` to use absolute paths
   - Updated gallery and map page routes to `/crowdnewsroom/[id]/gallery` and `/crowdnewsroom/[id]/map`
   - Updated typed router definitions to reflect new route structure

9. **Geocoding Integration**
   - Replaced legacy geocoding utilities with new `AddressFormatter`
   - Updated `reverseGeocode` to return `UnifiedAddress` instead of `GeocodeResult`
   - Integrated address formatting in map click handlers
   - Added support for address pattern formatting in callout responses

10. **Address Component Types**
    - Added `ADDRESS_COMPONENT_TYPE` enum for standardized address parsing
    - Added `UnifiedAddress` interface for consistent address data structure
    - Added `geocodeCountries` property to `CalloutMapSchema` for country restrictions

11. **Testing Framework**
    - Added Vitest as testing framework for frontend package
    - Created comprehensive test suite for `AddressFormatter`
    - Added test data from real geocoding results (CERN, NASA, Stonehenge)

### Technical Details

- **Address Provider**: Uses MapTiler's geocoding API with configurable language support and result filtering
- **Storage Provider**: Maintains existing Beabee storage implementation
- **Form Integration**: Automatically applies provider settings to prevent configuration errors
- **Error Prevention**: Hidden UI elements prevent users from selecting unsupported providers
- **Data Migration**: Existing forms are automatically updated to use correct providers
- **Address Formatting**: Supports placeholders like `{street_number} {address}, {postal_code} {municipality}` with fallback patterns
- **Testing**: Comprehensive test coverage using Vitest with real-world geocoding data

### Future Considerations

- Country restriction functionality for MapTiler provider (search for `geocodeCountries`)
- Additional provider support can be added by extending the base classes
- Provider selection UI can be re-enabled if multiple providers are needed

Based on the attached files, here's a brief description of the available placeholders:

### Available Address Placeholders

The address formatting supports the following placeholder types, derived from the `ADDRESS_COMPONENT_TYPE` enum:

### Basic Placeholders:
- `{street_number}` - House number (e.g. "90")
- `{address}` - Street name (e.g. "Hermannstraße") 
- `{postal_code}` - Postal code (e.g. "12051")
- `{country}` - Country (e.g. "Germany")

### Geographic Levels:
- `{municipality}` - Municipality/City (e.g. "Berlin")
- `{locality}` - District (e.g. "Mitte" in Berlin)
- `{region}` - Region/State (e.g. "Berlin")
- `{county}` - County (e.g. "Cuxhaven")
- `{subregion}` - Subregion (e.g. "Nordrhein" in NRW)

### Special Area Types:
- `{joint_municipality}` - Joint municipality
- `{municipal_district}` - Municipal district (e.g. "Mitte" in Berlin)
- `{neighbourhood}` - Neighbourhood (e.g. "Kreuzberg")
- `{place}` - Named place (e.g. "Brandenburger Tor")

### Usage:
- **Simple**: `{street_number} {address}` → "90 Hermannstraße"
- **With fallbacks**: `{municipality|county|region}` → uses first available component
- **Complete**: `{address} {street_number}, {postal_code} {municipality}` → "Hermannstraße 90, 12051 Berlin"

### Screenshots
[Video demonstration will be added here - showing the form builder with hidden provider options and address lookup functionality](https://github.com/user-attachments/assets/72c2f569-053c-4d3c-8e30-607c7244096e)

## Test
- [x] When a new upload form is created in a callout, provider selection should no longer be necessary and is now hidden
- [x] The same applies to addresses; autocomplete should work automatically
- [x] Address formatting should work as usual + the new one
- [x] Map should work as usual, search for place, set marker, etc

## Checklist before requesting a review

- [x] Done a self-review of my code
- [x] Run `yarn check` and addressed any problems
- [x] PR doesn't have merge conflicts

## Checklist before merging

- [x] Translations for all new i18n strings
- [x] Video demonstration added to PR description

## Ticket references:
- https://correctivdigital.openproject.com/wp/1058
- https://correctivdigital.openproject.com/wp/579
- https://correctivdigital.openproject.com/wp/1059
- https://correctivdigital.openproject.com/wp/1280

JumpLink committed:
docs(frontend): add comprehensive MapTiler integration documentation

- Add MapTiler integration guide with click-to-address workflow
- Add MapTiler address provider documentation for Form.io
- Add architecture decision record for MapTiler integration
- Include configuration, API usage, and troubleshooting guides
- Focus documentation structure on MapTiler-specific components

JumpLink committed:
feat(frontend): implement MapTiler address provider for Form.io

- Add MapTilerAddressProvider with full TypeScript support
- Integrate with existing MapTiler client configuration
- Support callout-specific country restrictions and current locale
- Provide createMapTilerProviderForCallout helper function
- Transform MapTiler responses to Form.io compatible format
- Register provider in Form.io providers system

JumpLink committed:
feat(frontend): enhance Form.io address provider integration

- Override default storage provider with BeabeeStorage
- Add MapTilerAddressProvider while maintaining default providers for backward compatibility
- Ensure seamless integration of new address provider into existing system

JumpLink committed:
docs(frontend): update MapTiler integration documentation

- Refine selection criteria for MapTiler, emphasizing open-source friendliness and cost-effectiveness
- Streamline configuration examples by removing unnecessary properties
- Clarify API rate limits and map loading behavior
- Remove outdated caching strategy and future enhancement suggestions for clarity

JumpLink committed:
fix(frontend): fix MapTiler provider API types and simplify implementation

- Fix invalid API type 'city' -> 'locality' to resolve 400 errors
- Remove unnecessary alternativeDisplayValueProperty complexity
- Simplify getDisplayValue method to use only place_name
- Remove complex attachAutocomplete stub implementation
- Provider now works correctly with MapTiler Geocoding API

JumpLink committed:
docs(frontend): update MapTiler provider documentation

- Fix API types from 'city' to 'locality' in examples
- Update attachAutocomplete documentation to reflect simplified implementation
- Add migration notes for v2.0.0 breaking changes
- Remove outdated complex autocomplete examples
- Clarify that Form.io handles autocomplete UI automatically

JumpLink committed:
docs(frontend): remove references to deleted attachAutocomplete method

- Remove documentation for attachAutocomplete method that was completely removed from provider
- Simplify autocomplete section to focus on Form.io's automatic handling
- Clean up migration notes to only mention removed features

JumpLink committed:
fix(frontend): update MapTilerAddressProvider to use current locale for language setting

- Change language parameter to always use the current language from reactive locale config
- Refactor return statement to ensure proper typing with FormioAddressResult
- Remove unused apiKey property from MapTilerSearchOptions interface

JumpLink committed:
fix(frontend): fix generate:index script

JumpLink committed:
feat(frontend): enhance MapTiler address provider functionality

- Introduce support for only the MapTiler address provider in the form builder
- Update BaseAddressProvider to improve option merging and type safety
- Refactor MapTilerAddressProvider to ensure proper request handling and parameter management
- Renamed FormioAddressProviderOptions interface
- Clean up and streamline MapTiler provider options and result structures

JumpLink committed:
docs(frontend): update MapTiler integration documentation

- Revise overview to clarify the focus on MapTiler integration and form builder
- Add new Form Builder documentation detailing dynamic form creation with MapTiler integration
- Streamline existing documentation by removing outdated sections and enhancing clarity
- Update references to environment configuration and prerequisites for better guidance

JumpLink committed:
feat(frontend): Set beabee storage provider as default

JumpLink committed:
feat(migrations): add migration to update Callout form schema components

- Implement migration to update storage and provider settings for form schema components in Callout
- Set storage to "beabee" for file components and provider to "maptiler" for address components
- Include logging for migration progress and updates

JumpLink committed:
docs(frontend): reduced documentation

JumpLink committed:
style(frontend): update form builder CSS to hide storage and provider elements

JumpLink committed:
Merge branch 'main' into feat/formio-address-provider

JumpLink committed:
docs(common): add geocodeCountries property documentation

wpf500 commented on apps/frontend/src/components/form-builder/form-builder.interface.ts:
```diff
@@ -174,6 +176,8 @@ export const formOpts = {
           schema: {
             type: 'file',
+            // We only support the beabee storage provider for file uploads
+            storage: 'beabee',
```
Haha so simple 🤦

wpf500 commented on docs/frontend/maptiler-integration.md:
Is this helpful documentation? I think it would be very easy for the code to get out-of-date without any real benefit

wpf500 commented on packages/core/src/migrations/1756466495968-UpdateCalloutFormSchemaComponents.ts:
I think this should be a manual migration rather than an automated database one, i.e. maybe as a CLI command like the image migration.

This could break a callout if anyone has managed to define a custom provider. I think this is unlikely and it's great that we are removing this option going forwards, but I would still want to check this per-client before rolling the change out.

wpf500 replied:
I don't think we should maintain backwards compatibility, but I do think we should know before we release which callouts could be affected so we can get in touch with the clients.

I wanted to avoid holding the release back by making the change opt-in, but actually now I think about it your automated migration is the best approach, and instead we can just temporarily pin the versions of the clients it could affect.

I agree we could use something that helps us easily run these sorts of reports, at the moment I use a script on the server to do it, but it's not helpful for others. For this release I suggest I use the old school method, and put this in our backlog together with the project of generally improving our infrastructure.

wpf500 commented on docs/frontend/maptiler-provider.md:
Is this documentation helpful? Shall we should discuss what sort of documentation we want in `docs/`? I would see this as being more useful as comments next to the relevant code.

wpf500 commented on apps/frontend/src/type/maptiler.ts:
Always so great how you improve the types as you go 🎉

wpf500 commented:
These are some really great changes in general by the way, you always push the Form.IO stuff in a much better direction (like previously creating all the types for the different component answer schemas etc.).

Also cool to see that the address lookup "just works", but we'll need to consider the consequences of this a bit more before releasing.

JumpLink committed:
docs(frontend): remove poorly structured MapTiler documentation

JumpLink committed:
feat(address): implement unified address data structures and formatter

- Add UnifiedAddress interface and AddressComponent types to common package
- Implement AddressFormatter class with transformation methods for MapTiler, Form.io, and backend
- Create MapTilerAddressProvider with Form.io integration using centralized formatter
- Update geocode utilities to use new AddressFormatter for consistent transformations
- Add ADDRESS_COMPONENT_TYPE enum for standardized address component types
- Migrate existing address formatting logic to use unified data structures

This implements Phase 1 of the Address Formatting Implementation Plan:
- Unifies data structures across MapTiler, Form.io, and backend formats
- Provides centralized address transformation and formatting
- Enables consistent address handling throughout the application

JumpLink committed:
chore(address): Renamed files to match the naming convention

JumpLink committed:
test: add vitest test runner and comprehensive AddressFormatter tests

JumpLink committed:
Merge pull request #326 from beabee-communityrm/feat/unify-geocode

feat(frontend/geocoding): Unify address data structures and centralize geocoding utilities

JumpLink committed:
Merge branch 'main' into feat/formio-address-provider

JumpLink committed:
refactor(tests): remove outdated comment from address formatter test file

wpf500 committed:
feat(address): unify geocode system and enhance address formatting

- Simplify fromFormio and toFormio by using unified address types directly
- Streamline map.vue and geocode.ts using AddressFormatter for cleaner code
- Add comprehensive ADDRESS_COMPONENT_TYPE support for more placeholder types

wpf500 reviewed (changes requested):
Some changes needed I think. We are really aiming for `CalloutResponseAnswerAddress` to be _the_ unified response address schema. I'm going to have a go at this

wpf500 committed:
refactor(address): Remove intermediary address types

This commit removes all intermediary address types in favour of
CalloutResponseAddressAnswer. This type represents the underyling
data structure in the database, we should be converting all other
geocoding data formats into this.

This commit introduces some backwards incompatible changes to the
CalloutResponseAddressAnswer type which will need to be addressed with a
database migration.

wpf500 committed:
chore(map): just store click location, no need to parse everything

wpf500 committed:
chore(map): revert check for location, it should always exist

wpf500 committed:
chore(geocode): return all geocode features instead of just a subset

We should store as much about the geocoded feature as possible, in the
future the address template could be updated and we might need different
data

wpf500 committed:
feat(address): support addresses without street level data

wpf500 committed:
feat(address): make new address format importable

wpf500 committed:
chore(anonymiser): update address anonymiser with new structure

wpf500 committed:
fix(maptiler-provider): use correct display value property

wpf500 committed:
chore(form-builder): hide unused address field configuration

wpf500 committed:
feat(migration): add migration for responses with an address in them

RayMille reviewed (dismissed):
To be honest, I find it a bit challenging to review this since there are many parts of the code I haven’t worked on yet. From my perspective, though, it looks good—I don’t see anything obviously problematic. I also ran it on my machine, and it works :) :tada:

wpf500 commented:
This is ready, but because of the more complicated data migration we'll need to do a specific release for it with a few additional steps, so I'm leaving this unmerged for now.

JumpLink committed:
Merge branch 'main' into feat/formio-address-provider

JumpLink committed:
Merge branch 'main' into feat/formio-address-provider

JumpLink committed:
chore: Regenerate index files

JumpLink committed:
Merge branch 'main' into feat/formio-address-provider

JumpLink committed:
Merge branch 'main' into feat/formio-address-provider

JumpLink committed:
fix(backend-cli): add missing fields to imported address answers

JumpLink committed:
fix(migration): add missing source field to address answer migration

The target type and converted answer object were both missing source:
'maptiler', which is required by CalloutResponseAnswerAddress.

JumpLink committed:
fix(migration): use component key instead of id for answer lookup

Answers are stored in callout_response.answers by component.key
(e.g. "addressHidden"), not by component.id (e.g. "ewtg3cn").
Using id caused all address answers to be silently skipped.

JumpLink committed:
fix(migration): rewrite address conversion using native JSONB SQL

TypeORM queryRunner.query() returns JSONB as non-mutable JS objects,
so mutations were silently lost before JSON.stringify. Replace the
JS fetch/mutate/write pattern with a single SQL UPDATE using
jsonb_set + jsonb_build_object to transform answers in-database.

JumpLink committed:
fix(migration): parse JSONB formSchema string returned by queryRunner

TypeORM's queryRunner.query() may return JSONB columns as raw strings
rather than parsed objects. Without JSON.parse the formSchema.slides
access is undefined, causing addressComponents to be empty and skipping
all response migrations silently.

JumpLink committed:
Merge branch 'main' into feat/formio-address-provider

wpf500 commented on apps/frontend/src/lib/index.ts:
```diff
@@ -1,6 +1,9 @@
+export * from './address.formatter.test-data';
```
This shouldn't be exporting test-data files. I guess this is auto generated and maybe needs an additional rule to exclude test-data files?

JumpLink replied:
Thank you, good catch, fixed in https://github.com/beabee-communityrm/monorepo/pull/323/commits/9b634ad1610a55827147b265c556b4854b50d5dc

wpf500 reviewed (changes requested):
Last quick change before merging

JumpLink committed:
fix(dev-cli): exclude test-data files from generated indexes

Prevents test-data modules from being re-exported via auto-generated
index.ts files. Regenerates apps/frontend/src/lib/index.ts to drop the
stale address.formatter.test-data export.
